### PR TITLE
Merge derivation into derive and remove lift

### LINKED ISSUE #61: Merge `derivation` into `derive` and remove `lift`
At the moment the `derive` function is only used for tagged template strings. I think it should still be used for that, but that it should also be used for the same purpose as `derivation` currently is, and in some ways should mirror the `.derive` method in accepting extra arguments.

e.g.

- `x.derive(f)` should be equivalent to `derive(f, x)` should be equivalent to `derive(() => f(x.get()))`
- `x.derive(f, y)` should be equivalent to `derive(f, x, y)` should be equivalent to `derive(() => f(x.get(), y.get()))`

and so on.

w/r/t/ replacing `lift`:

`lift(f)(x, y)` now becomes `derive(f, x, y)` and `lift(f)` becomes `derive.bind(null, f)` or equivalent, if you're using some nifty library with partial function application.

### PR DESCRIPTION
**Pull request by oskbor**
I had to move the unpack method since I got cyclic dependencies when I imported it into `derivation`from `module`. Let me know where you want me to put the `unpack`method.

**Remaining**
 - [x] Where to put `unpack`
 - [x] Tests for the new signatures
 - [x] Update flow and TypeScript definitions
 - [x] Update the CHANGELOG

Closes #61

### CODE REVIEW

File: src/derivation.js [hunk 1]
```diff
@@ -121,3 +122,52 @@ export function detach (parent, child) {
 export function derivation (deriver) {
   return new Derivation(deriver);
 }
+
+export function derive (f, a, b, c, d) {
+  if (f instanceof Array) {
+    // Template string tag for derivable strings
+    var args = util.slice(arguments, 1);
+    return derivation(function () {
```
Comment (oskbor) [inline src/derivation.js]: should I perhaps return a `new Derivation()`s instead?

Comment (ds300) [inline src/derivation.js]: Yeah why not. Always good to keep stack traces short.

File: src/unpack.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import {isDerivable} from './types';
+import * as util from './util';
+
+/**
+ * dereferences a thing if it is dereferencable, otherwise just returns it.
+ */
+
+export function unpack (thing) {
```
Comment (ds300) [inline src/unpack.js]: I think it is fine to keep `unpack` in a separate file to avoid cyclical dependencies.

**Comment by oskbor** [comment]
I am not sure I got the typescript and flow typings totally correct, so you might want to take a look at that :slightly_smiling_face:

**Comment by ds300** [comment]
Cool cool cool :+1: Looks like there are still a few usages of `lift`, in the examples and the tests.

**Comment by ds300** [comment]
Oh silly me, you just redefined `lift` in terms of `derive` for the test. I didn't read through it properly earlier.

Let's still fix up the examples and make them pull in the `derivable` package from `"file:../../../"` rather than npm, so they don't break as we keep them up to date with master.

**Comment by oskbor** [comment]
Not sure on how to get coverage up again,
and it seems most or all examples are broken :slightly_frowning_face: 
Will start with fixing the examples

**Comment by oskbor** [comment]
webpack fails in the react examples with 
```
ERROR in ./~/react-derivable/lib/index.js
Module build failed: Error: Couldn't find preset "prometheusresearch" relative to directory "/home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/react-derivable"
    at /home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/babel-core/lib/transformation/file/options/option-manager.js:293:19
    at Array.map (native)
    at OptionManager.resolvePresets (/home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/babel-core/lib/transformation/file/options/option-manager.js:275:20)
    at OptionManager.mergePresets (/home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/babel-core/lib/transformation/file/options/option-manager.js:264:10)
    at OptionManager.mergeOptions (/home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/babel-core/lib/transformation/file/options/option-manager.js:249:14)
    at OptionManager.init (/home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/babel-core/lib/transformation/file/options/option-manager.js:368:12)
    at File.initOptions (/home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/babel-core/lib/transformation/file/index.js:212:65)
    at new File (/home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/babel-core/lib/transformation/file/index.js:135:24)
    at Pipeline.transform (/home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/babel-core/lib/transformation/pipeline.js:46:16)
    at transpile (/home/oskbor/Projects/derivablejs/examples/bmi-calculator-advanced/react/node_modules/babel-loader/lib/index.js:46:20)
 @ ./main.js 15:22-48
```
Perhaps does @andreypopp have some clues? (error seems to originate in `react-derivable`)

**Comment by andreypopp** [comment]
Looks like webpack is configured to apply Babel on files found in
node_modules which is not needed.

**Comment by ds300** [comment]
Thanks for all your work on this :) Don't worry about the coverage thing. I'll configure coveralls to be less strict.

File: examples/bmi-calculator-advanced/react/main.js [hunk 3]
```diff
@@ -1,9 +1,16 @@
-import {atom, lift, transaction} from 'derivable';
+import {atom, derive, transaction} from 'derivable';
 import React from 'react';
 import {render} from 'react-dom';
 import {reactive} from 'react-derivable';
 
 /**
+ * Define lift as a function of derive
+ */
+function lift (fn) {
```
Comment (oskbor) [inline examples/bmi-calculator-advanced/react/main.js]: ah, yes thanks for catching this :+1:

**Pull request merged**